src/oracle/monitor.js

```diff
@@ -0,0 +1,429 @@
+        if (id <= lastOfferId) {
+            break
+        }
+
+        //if a price invitation
+        if (currentOffer["status"]["invitationSpec"]["invitationMakerName"] == "PushPrice") {
+
+            let feed = feeds[currentOffer["status"]["invitationSpec"]["previousOffer"]]
+            let lastRound = Number(currentOffer["status"]["invitationSpec"]["invitationArgs"][0]["roundId"])
+
+            //get feeds last observed round from state
+            let lastObservedRound = state["values"].hasOwnProperty(feed) ? state["values"][feed]["round"] : 0
+
+            //if round is bigger than last observed and the offer didn't fail
+            if (lastRound > lastObservedRound && !currentOffer["status"].hasOwnProperty("error")) {
+
+                //if id is bigger than last offer id in state, set it
+                lastResults["last_offer_id"] = id
+
+                let price = Number(currentOffer["status"]["invitationSpec"]["invitationArgs"][0]["unitPrice"]) / amountsIn[feed]
+
+                //fill results variable
+                lastResults["values"][feed] = {
+                    price: price,
+                    id: id,
+                    round: lastRound
+                }
+                state = lastResults
+
+                //get latest feed price
+                let feedPrice = await queryPrice(feed)
+                //update metrics
+                updateMetrics(oracleDetails["oracleName"], oracle, feed, price, id, feedPrice, lastRound)
+
+            }
+        }
+    }
+
+    //loop through balances
+    for (var i = 0; i < offersBalances.balances.length; i++) {
+        let currentBalance = offersBalances.balances[i]
+
+        var brand = currentBalance.brand.iface.split(" ")[1]
+        if (brand.includes("BLD") || brand.includes("IST")) {
+            var value = Number(currentBalance.value)
+            updateBalanceMetrics(oracleDetails["oracleName"], oracle, brand, value)
+        }
+    }
+
+    return lastResults["last_offer_id"] != lastOfferId ? lastResults : state
+}
+
+/**
+ * Function to read the latest monitoring state from file
+ * @returns latest monitoring state
+ */
+const readMonitoringState = () => {
+    //try to read from file
+    try {
+        return readJSONFile(STATE_FILE)
+    } catch (err) {
+        //if it fails, initialise and save
+        let initialState = {}
+
+        for (let oracle in oracles) {
+            initialState[oracle] = {
+                "last_index": 0,
+                "values": {}
+            }
+        }
+
+        //save to file
+        saveJSONDataToFile(initialState, STATE_FILE)
+        return initialState
+    }
+}
+
+/**
+ * Main function to monitor
+ */
+export const monitor = async () => {
+
+    //create interval
+    setInterval(async () => {
+
+        //read monitoring state
+        let state = readMonitoringState()
+
+        //for each oracle
+        for (let oracle in oracles) {
```

Comment:
Shouldn't getOraclesInvitations() be done here instead of outside (or ran periodically) since the node operator needs to stop the terminal and re-run it to see new invitations

Reply:
Fixed in https://github.com/jacquesvcritien/agoric-cl-middleware/commit/4fc58332861c3253a32219b20396790a4159e06e